src/main/java/org/eclipse/uprotocol/communication/RpcClient.java

```diff
@@ -0,0 +1,123 @@
+ * 
+ * clients use this API to invoke a method (send a request and wait for a reply).
+ */
+public interface RpcClient {
+
+    /**
+     * API for clients to invoke a method (send an RPC request) and receive the response which is 
+     * a {@link CompletionStage} for the response payload. <br>
+     * 
+     * @param transport The transport to send the request over.
+     * @param methodUri The method URI to be invoked.
+     * @param message The optional request message to be sent to the server.
+     * @return Returns the CompletionStage with the response payload or exception with the failure
+     * reason as {@link UStatus}.
+     */
+    static CompletionStage<UPayload> invokeMethod(UTransport transport, UUri methodUri, Message message) {
+        return invokeMethod(transport, methodUri, UPayload.pack(message), CallOptions.DEFAULT);
+    }
+
+
+    /**
+     * API for clients to invoke a method (send an RPC request) and receive the response which is 
+     * a {@link CompletionStage} for the response payload. <br>
+     * 
+     * @param transport The transport to send the request over.
+     * @param methodUri The method URI to be invoked.
+     * @param requestPayload The request message to be sent to the server.
+     * @return Returns the CompletionStage with the response payload or exception with the failure
+     * reason as {@link UStatus}.
+     */
+    static CompletionStage<UPayload> invokeMethod(UTransport transport, UUri methodUri, UPayload requestPayload) {
+        return invokeMethod(transport, methodUri, requestPayload, CallOptions.DEFAULT);
+    }
+    
+    
+    /**
+     * API for clients to invoke a method (send an RPC request) and receive the response (the returned 
+     * {@link CompletionStage} {@link UPayload}. <br>
+     * 
+     * @param transport The transport to send the request over.
+     * @param methodUri The method URI to be invoked.
+     * @param requestPayload The request message to be sent to the server.
+     * @param options RPC method invocation call options, see {@link CallOptions}
+     * @return Returns the CompletionStage with the response payload or exception with the failure
+     * reason as {@link UStatus}.
+     */
+    static CompletionStage<UPayload> invokeMethod(UTransport transport, UUri methodUri, UPayload requestPayload, CallOptions options) {
+        Objects.requireNonNull(transport, UTransport.TRANSPORT_NULL_ERROR);
+        UMessageBuilder builder = UMessageBuilder.request(transport.getSource(), methodUri, options.timeout());
+        UMessage request;
+        
+        options.token().ifPresent(builder::withToken);
+
+        request = requestPayload == null ? builder.build() : builder.build(requestPayload.getFormat(), requestPayload.getData());
+        
+        CompletionStage<UPayload> responsePayload = new CompletableFuture<UPayload>();
+        
+        transport.registerListener(methodUri, transport.getSource(), new UListener() {
```

Comment:
@mishap4 ,  OK I addressed this in the latest change by filtering in the listener by the reqid. This allows me to not have to introduce state and/or instance in L2. If this turns out to be a performance issue (because someone invokes the same 
message for a thousand times in a row) then we will reevaluate the situation.

Reply:
L2 changes have been pushed out to #117